design/proposals/module-linking/Explainer.md

```diff
@@ -841,34 +845,60 @@ binaries in all the same places where core module binaries can be loaded today,
+
+For adapter module imports of module type, we need a fundamentally new way to
+request that the ESM loader parse or decode a module without *also*
+instantiating that module. Recognizing this same need from JavaScript, there is
+already a TC39 proposal called [Import Reflection] that adds the ability to
+write, in JavaScript:
+```js
+import Foo from "./foo.wasm" as "wasm-module";
+assert(Foo instanceof WebAssembly.Module);
+```
+With this extension to JavaScript and the ESM loader, an adapter module import
+of module type can be treated like an `as "wasm-module"` import by
+ESM-integration.
 
-For single-level imports of non-instance type, the natural analogue in an ESM
-context is the [default export]. Thus, a single-level non-instance import would
-receive whatever the ESM loader says the default value would be.
+In all other cases, the (single) string imported by an adapter module import is
+first resolved to a [Module Record] using the same process as resolving the
+[Module Specifier] of a JavaScript `import`. After this, the handling of the
+imported Module Record is determined by the import type:
 
 For imports of instance type, the ESM loader would treat the exports of the
-instance type as the imported field names. In the normal two-level
-import-an-instance case, this corresponds exactly to existing ESM-integration
-behavior. However, for more-deeply nested instances, the ESM loader would need
-to recursively extract fields.
-
-Lastly, for imports of modules, ESM-integration would need the ESM loader to
-provide a fundamentally new way to parse/decode a module without *also*
-instantiating that module. Since this functionality is also useful for JS,
-there is already a proposal to add this to ESM called [Import Reflection].
-With this proposal, a module import:
+instance type as if they were the [Named Imports] of a JavaScript `import`.
+Thus, single-level imports of instance type mirror the behavior of core
+two-level imports with existing ESM-integration. Since the exports of an
+instance type can themselves be instance types, this process must be performed
+recursively.
+
+For imports of type `(global externref)` (i.e., an immutable core WebAssembly
+global storing a value of type `externref`), the imported Module Record is
+first converted to a [Namespace Object] via [`GetModuleNamespace()`] and then
+converted to a `(global externref)` as a normal JavaScript value. This allows
+the importing adapter module to dynamically access the fields of the
+Namespace Object by calling other imported functions (such as `Object.get()`).
+In the future, other compatible types (e.g., `handle` Interface Types) could be
+included in this case.
+
+Otherwise, if the import cannot accept a Namespace Object (and would error
+otherwise), the import is treated like a JavaScript [Imported Default Binding]
+and the Module Record is converted to its default value. This allows, for
+example, a single level import of a function:
```

Comment:
Once concern here is that the WebAssembly ESM integration currently specifies itself on top of the JS API. That is, we are implying that `(import "./foo.js" (func (result i32)))` would be generated by `importObj = { './foo.js': f }` in the JS API, so that the ESM integration is having to do type reflection to determine whether to pass the namespace or the direct function to the `importObj` of the WebAssembly JS API. Will this reflection ability itself be exposed to JS to be able to ensure it can be done predictably from JS?

As I've said before, one way to make this more well-defined is to treat `importObj: { './foo.js': f }` as sugar for `importObj: { './foo.js': { default: f } }` such that the single / two-level differences are fully encapsulated into the JS API and adapter module semantics, instead of relying on outside inference on the boundaries. This might help interface predictability.

So the suggestion would be:
1. Have the default access be encoded into the adapter module JS API integration logic itself
2. Also in the JS API, treat single level imports as sugar for `{ default: value }`

The benefit of the above is then that `(import './foo.js' (func (result i32)))` could always be populated through the ESM API via namespace values without needing to perform reflection on the Web Assembly binary to determine the import type rules, which might be a source of bugs otherwise.

Of course outside of JS none of the above are necessary - one could just go direct without needing this indirection, and the direct attachment can be performed, but so far as adapter modules interface with the JS API, treating the JS API as the place where this interop is handled may help make things a little more predictable possibly to ensure a more direct correspondence between the JS API and ESM integration without reflection.

Unless I've missed a place where this could conflict otherwise?

Reply:
Hmm, I see your point about it making sense to try to push this down into the JS API, but I worry that having the JS API contain an ad hoc rule for `default` will end up feeling pretty irregular in the context of the JS API and it could end up surprising someone in a bad way.

Actually, thinking about this more, I keep feeling uncomfortable about the special-case of the `global externref` type -- this seems fragile in a way that I worry may also lead to surprising behavior as the reference type hierarchy grows in complexity.  I wonder if instead we should capture the intention to import a Namespace Object in the same manner as we talked about JSON/CSS modules: using a type import:
```wasm
(adapter module
  (import "std:types:Namespace" (type $NamespaceObject))
  (import "foo" (global (ref $NamespaceObject)))
)
```
That way we're being super-explicit about what we are wanting to do here.  This would also mean that this feature wouldn't be in the initial release of Module Linking + ESM-integration since we don't have type imports or built-in modules, but as I think you pointed out in the original issue: this doesn't seem like a super-high-priority use case, just one we should retain the option of supporting it later.  So making the "MVP" simpler is I think a win.

Based on this and getting to your original point about ESM-integration needing to type-reflection to determine which value to pass down to the JS API instantiation logic: given that the ESM loader will also have to inspect the imports' types to determine how to materialize the JSON and CSS module import values (inspecting type import names, etc), this seems fine and symmetric for Namespace Objects.

Thus, my concrete proposed change is: remove the Namespace Object special case for `global externref` and unconditionally perform default-conversion for all non-module, non-instance single-level imports. WDYT?

Reply:
That does seem to simplify the model, it would just be important to have the associated reflection via the JS API to be able to inspect these types, while the model I suggested above wouldn't have to rely on that as the main feature.

Is there any such reflection work in progress anywhere?

Can you perhaps clarify what makes you feel uneasy about that model a little? Of course the assertions / reflection type requires reflection, but the invariant today is all Wasm module imports can be blindly populated by _any_ module namespace (ie the result of a dynamic import resolution). Guarding this on a conditional population adds a point of friction in interop which just needs to be done carefully. Eg replacing a Wasm module with a JS module becomes lossy in this process and isn't a simple replacement (although we already discussed `export default wasmModule` perhaps not being necessary, it's still a loss of generality / transposability).

Reply:
Oh, the only thing I was uneasy about is having the JS API do the default-conversion of Namespace Objects automatically in the definition of (presumably) [read the imports](https://webassembly.github.io/spec/js-api/index.html#read-the-imports).

For reflection, I think, as you suggested separately, we can use `WebAssembly.exports()`, which will be extended to have `module` and `instance` import `kind`s.  So then I'd suggest that ESM-integration would reflect on the import types, and for non-module/instance imports, perform default conversion to produce the JS value passed to *read the imports*.  How does that sound?

Reply:
Got it, thanks for taking the time to go through this. The reflection is just a necessary step then always for arbitrary Wasm linking, but since it integrates well into the existing mechanism anyway it's a clear integration so I don't think there's any confusion so long as we avoid too many custom type mapping rules / or at least make those rules very clear and well defined.

Reply:
(Also, I added a note to the JS API section about `imports()`/`exports()`.)